src/main/java/org/jabref/gui/importer/fetcher/WebSearchPaneViewModel.java

```diff
@@ -91,11 +109,62 @@ public void search() {
 
         BackgroundTask<ParserResult> task;
         task = BackgroundTask.wrap(() -> new ParserResult(activeFetcher.performSearch(getQuery().trim())))
-                             .withInitialMessage(Localization.lang("Processing %0", getQuery().trim()));
+                .withInitialMessage(Localization.lang("Processing %0", getQuery().trim()));
         task.onFailure(dialogService::showErrorDialogAndWait);
 
         ImportEntriesDialog dialog = new ImportEntriesDialog(stateManager.getActiveDatabase().get(), task);
         dialog.setTitle(activeFetcher.getName());
         dialogService.showCustomDialogAndWait(dialog);
     }
+
+    public void validateQueryStringAndGiveColorFeedback(TextField querySource, String queryString) {
+        if (!queryString.strip().isBlank() && !isPatternMatched(queryPattern, queryString)) {
+            if (isPatternMatched(laxQueryPattern, queryString)) {
+                setPseudoClassToUnsupported(querySource);
+                querySource.setTooltip(new Tooltip(Localization.lang("This query uses unsupported fields.")));
+            } else {
+                setPseudoClassToInvalid(querySource);
+                querySource.setTooltip(new Tooltip(Localization.lang("This query uses unsupported syntax.")));
+            }
+        } else if (containsYearAndYearRange(queryString)) {
+            setPseudoClassToInvalid(querySource);
+            querySource.setTooltip(new Tooltip(Localization.lang("The query cannot contain a year and year-range field.")));
+        } else {
+            setPseudoClassToValid(querySource);
+        }
+    }
+
+    private void setPseudoClassToUnsupported(TextField querySource) {
+        querySource.pseudoClassStateChanged(PseudoClass.getPseudoClass("invalid"), false);
+        querySource.pseudoClassStateChanged(PseudoClass.getPseudoClass("unsupported"), true);
+    }
+
+    public void setPseudoClassToValid(TextField querySource) {
+        querySource.pseudoClassStateChanged(PseudoClass.getPseudoClass("invalid"), false);
+        querySource.pseudoClassStateChanged(PseudoClass.getPseudoClass("unsupported"), false);
+    }
+
+    private void setPseudoClassToInvalid(TextField querySource) {
+        querySource.pseudoClassStateChanged(PseudoClass.getPseudoClass("invalid"), true);
+    }
+
+    private boolean containsYearAndYearRange(String queryString) {
+        return queryString.toLowerCase().contains("year:") && queryString.toLowerCase().contains("year-range:");
+    }
+
+    // Use Thread Pool to control the match time
+    private boolean isPatternMatched(Pattern pattern, String queryString) {
+        boolean isMatched = true;
+
+        try {
+            BackgroundTask.wrap(() -> pattern.matcher(queryString.strip()).matches())
```

Comment:
In this case, it's actually function `matches()` takes a lot of time that make GUI freeze. When we input a long title, it will take about 2 mintues, which will make GUI freeze

Reply:
oooh that's surprising. The query parser is quicker, so this shouldn't be a problem any longer once use it.

Reply:
I am very sorry to hear this.....since I need to merge a certain number PR due to some personal reasons....
Actually, I indeed spent a lot of time dealing with this bug.

But in this milestone, I think this PR actually solve the problem of GUI freeze. If possible, I hope you can merge this request temparaily, and then consider further solution in the next milestone together!

Anyway, thanks to all developers in your team. I learned a lot from your fast responses and right guides.

Reply:
If you use the query parser instead of a regex, then it's good to go from my side.